userchroot.c

```diff
@@ -313,6 +314,33 @@ static int child_fn(void* v) {
   return 0;
 }
 
+static int proc_guard(void *v) {
+
+    pid_t child_pid =
+        clone(
+            child_fn,
+            child_stack+sizeof(child_stack),
+            CLONE_NEWNS | CLONE_NEWPID | SIGCHLD,
+            v
+        );
+    if(-1 == child_pid) {
+      fprintf(stderr, "Failed to clone. Error: %s\n", strerror(errno));
+    }
+    else {
+        int child_status = 0;
+        waitpid(child_pid, &child_status, 0);
```

Comment:
I realize now that this waitpid and the other one also need a small loop around it. If the process is interrupted by any signal, waitpid may return 0. So you need to do a loop with waitpid to make sure that the pid you wanted was returned.